src/Raven.Client/Documents/Session/IAsyncDocumentQuery.cs

```diff
@@ -67,6 +67,13 @@ public interface IAsyncRawDocumentQuery<T> :
     {
     }
 
+    public interface IAsyncGraphQuery<T> :
+        IQueryBase<T, IAsyncGraphQuery<T>>,
+        IAsyncDocumentQueryBase<T>
+    {
+        IAsyncGraphQuery<T> With<TOther>(string alias, IRavenQueryable<TOther> query, CancellationToken token = default);
```

Comment:
Let's remember to add the overload with `IDocumentQuery<TOther>`

Reply:
I need to see if my solution works for IDocumentQuery since the tokens might be already populated with parameter names

Reply:
I might need to transform the DocumentQuery to a queryable and back so i could modify the tokens. writing a test to check this senario.

Reply:
@ayende @ppekrol 
what if we have something like so:
.With("Foo", session.Advanced.AsyncDocumentQuery<Foo>(), q => q.WhereIn(x=>x.Name, names))
The issue here is that i need a clean DocumentQuery, i could replace the empty AsyncDocumentQuery with those three parameters:
string indexName , string collectionName , bool isMapReduce
to construct the query myself but if not then we can't use my approach and we are back to iterating the tokens

Reply:
We can either drop the support for IDocumentQuery or do:
```
.With("Foo",q => q.DocumentQuery<TIndex>().WhereIn(x=>x.Name, names));
```

The q will be our own 'builder' that will do:
```
public IFilterDocumentQuery<T> DocumentQuery<TIndex>()
{
    var dq = session.Advanced.DocumentQuery<TIndex>();
    dq.PrefixParameters = "abc";
    return dq;
}
```